File: sample_code/cplusplus/Tetris/Block.cpp

#include "Block.h"

Block Block::moveDown() {
    Block b(*this);
    int i = 0;

    int __action__ = 1;

    while(true) {
        switch(__action__) {
            case 1:
            if(i < 8) {
                __action__ = 2;
            }
            else {
                __action__ = 3;
            }
            break;

            case 2:
            b.coordinates[i + 1] += 1;
            i += 2;
            __action__ = 1;
            break;

            case 3:
            return b;
        }
    }
}

int Block::getKind() {
    return kind;
}

int * Block::getCoordinates() {
    return coordinates;
}

Block Block::moveRight() {
    Block b(*this);
    int i = 0;

    int __action__ = 0;

    while(true) {
        switch(__action__) {
            case 0:
            if(i < 8) {
                __action__ = 2;
            }
            else {
                __action__ = 3;
            }
            break;

            case 2:
            b.coordinates[i] += 1;
            i += 2;
            __action__ = 0;
            break;

            case 3:
            return b;
        }
    }
}

Block::Block(int k) {
    int *coord, i = 0;
    static int Oc[] = {5, 1, 4, 1, 4, 2, 5, 2};
    static int Tc[] = {5, 1, 6, 1, 4, 1, 5, 2};
    static int Sc[] = {5, 1, 6, 1, 4, 2, 5, 2};
    static int Ic[] = {4, 2, 3, 2, 5, 2, 6, 2};
    static int Lc[] = {5, 1, 4, 1, 6, 1, 4, 2};
    static int Jc[] = {5, 1, 4, 1, 6, 1, 6, 2};
    static int Zc[] = {5, 1, 4, 1, 5, 2, 6, 2};
    static int *coords[] = {0, Tc, Sc, Ic, Lc, Jc, Zc, Oc};

    int __action__ = 2;

    while(true) {
        switch(__action__) {
            case 0:
            if(i < 8) {
                __action__ = 3;
            }
            else {
                __action__ = 4;
            }
            break;

            case 2:
            kind = k;
            coord = coords[k];
            __action__ = 0;
            break;

            case 3:
            coordinates[i] = *(coord + i);
            i++;
            __action__ = 0;
            break;

            case 4:
            return;
        }
    }
}

Block Block::rotate() {
    Block b(*this);
    int i = 2, xc = coordinates[0], yc = coordinates[1], s = xc + yc, d = xc - yc;

    int __action__ = 2;

    while(true) {
        switch(__action__) {
            case 1:
            if(i < 8) {
                __action__ = 3;
            }
            else {
                __action__ = 4;
            }
            break;

            case 2:
            if(b.kind != Block::O) {
                __action__ = 1;
            }
            else {
                __action__ = 4;
            }
            break;

            case 3:
            b.coordinates[i] = coordinates[i + 1] + d;
            b.coordinates[i + 1] = s - coordinates[i];
            i += 2;
            __action__ = 1;
            break;

            case 4:
            return b;
        }
    }
}

Block Block::moveLeft() {
    Block b(*this);
    int i = 0;

    int __action__ = 1;

    while(true) {
        switch(__action__) {
            case 1:
            if(i < 8) {
                __action__ = 2;
            }
            else {
                __action__ = 3;
            }
            break;

            case 2:
            b.coordinates[i] -= 1;
            i += 2;
            __action__ = 1;
            break;

            case 3:
            return b;;
        }
    }
}

Block::Block(const Block & b) {
    int i = 0;

    int __action__ = 3;

    while(true) {
        switch(__action__) {
            case 0:
            if(i < 8) {
                __action__ = 2;
            }
            else {
                __action__ = 4;
            }
            break;

            case 2:
            coordinates[i] = b.coordinates[i];
            i++;
            __action__ = 0;
            break;

            case 3:
            kind = b.kind;
            __action__ = 0;
            break;

            case 4:
            return;
        }
    }
}

File: sample_code/java/Tetris/Block.java

public class Block {
    public final static int L = 4;
    public final static int O = 7;
    public final static int T = 1;
    public final static int S = 2;
    public final static int Z = 6;
    public final static int J = 5;
    public final static int VOID = 0;
    public final static int I = 3;

    private int kind;
    private int [] coordinates;

    public Block moveDown() {
        Block b = new Block(this);
        int i = 0;

        int __action__ = 0;
        Block __returned__;

        __loop__: while(true) {
            switch(__action__) {
                case 0:
                if(i < 8) {
                    __action__ = 2;
                }
                else {
                    __action__ = 3;
                }
                break;

                case 2:
                b.coordinates[i + 1] += 1;
                i += 2;
                __action__ = 0;
                break;

                case 3:
                __returned__ = b;
                break __loop__;
            }
        }

        return __returned__;
    }

    public int getKind() {
        return kind;
    }

    public int [] getCoordinates() {
        return coordinates;
    }

    public Block moveRight() {
        Block b = new Block(this);
        int i = 0;

        int __action__ = 1;
        Block __returned__;

        __loop__: while(true) {
            switch(__action__) {
                case 1:
                if(i < 8) {
                    __action__ = 2;
                }
                else {
                    __action__ = 3;
                }
                break;

                case 2:
                b.coordinates[i] += 1;
                i += 2;
                __action__ = 1;
                break;

                case 3:
                __returned__ = b;
                break __loop__;
            }
        }

        return __returned__;
    }

    public Block(int k) {
        int [] coord;
        int [] Ocoord = {5, 1, 4, 1, 4, 2, 5, 2};
        int [] Tcoord = {5, 1, 6, 1, 4, 1, 5, 2};
        int [] Scoord = {5, 1, 6, 1, 4, 2, 5, 2};
        int [] Icoord = {4, 2, 3, 2, 5, 2, 6, 2};
        int [] Lcoord = {5, 1, 4, 1, 6, 1, 4, 2};
        int [] Jcoord = {5, 1, 4, 1, 6, 1, 6, 2};
        int [] Zcoord = {5, 1, 4, 1, 5, 2, 6, 2};

        kind = k;

        switch(k) {
            case O: coord = Ocoord; break;
            case T: coord = Tcoord; break;
            case S: coord = Scoord; break;
            case I: coord = Icoord; break;
            case L: coord = Lcoord; break;
            case J: coord = Jcoord; break;
            default: coord = Zcoord; break;
        }

        coordinates = new int[8];
        for(int i = 0; i < 8; i++)
        coordinates[i] = coord[i];
    }

    public Block rotate() {
        Block b = new Block(this);
        int i = 0, xc = coordinates[0], yc = coordinates[1], s = xc + yc, d = xc - yc;

        int __action__ = 2;
        Block __returned__;

        __loop__: while(true) {
            switch(__action__) {
                case 0:
                if(i < 8) {
                    __action__ = 3;
                }
                else {
                    __action__ = 4;
                }
                break;

                case 2:
                if(b.kind == Block.O) {
                    __action__ = 4;
                }
                else {
                    __action__ = 0;
                }
                break;

                case 3:
                b.coordinates[i] = coordinates[i + 1] + d;
                b.coordinates[i + 1] = s - coordinates[i];
                i += 2;
                __action__ = 0;
                break;

                case 4:
                __returned__ = b;
                break __loop__;
            }
        }

        return __returned__;
    }

    public Block moveLeft() {
        Block b = new Block(this);
        int i = 0;

        int __action__ = 1;
        Block __returned__;

        __loop__: while(true) {
            switch(__action__) {
                case 1:
                if(i < 8) {
                    __action__ = 2;
                }
                else {
                    __action__ = 3;
                }
                break;

                case 2:
                b.coordinates[i] -= 1;
                i += 2;
                __action__ = 1;
                break;

                case 3:
                __returned__ = b;
                break __loop__;
            }
        }

        return __returned__;
    }

    public Block(Block b) {
        int i = 0;

        int __action__ = 2;

        __loop__: while(true) {
            switch(__action__) {
                case 1:
                if(i < 8) {
                    __action__ = 3;
                }
                else {
                    __action__ = 4;
                }
                break;

                case 2:
                kind = b.kind;
                coordinates = new int[10];
                __action__ = 1;
                break;

                case 3:
                coordinates[i] = b.coordinates[i];
                i++;
                __action__ = 1;
                break;

                case 4:
                break __loop__;
            }
        }
    }
}
